### ISSUE TITLE
Significantly different CPU-Usage and Load Average after Update to 4.1

### ISSUE THREAD

**Issue by Cutuchiqueno** [ManifoldScholar/manifold#2703]
I recognized that CPU-Usage and Load Average changed significantly after migrating to Manifold 4.1 from 4.0. To better demonstrate this I prepared this picture:

![image](https://user-images.githubusercontent.com/306064/80600220-5bb4ef80-8a2c-11ea-8968-66233e47ae53.png)

So compared to before the update, there are constantly processes waiting in the queue and the general workload for the processor is now stable around 18%/20% (following htop) compared o 4-5% before.

I tried to track down in htop what is going on. I am not 100% sure, but the most demanding and frequently state changing process is a ruby script called `zhong.rb`.

Did the upgrade introduce changes that justify this change in workload?

**Comment by zdavis** [comment]
We're not seeing spikes this high on our instances, so it's probably worth debugging a bit. Zhong is what Manifold uses for running scheduled tasks. I believe that if you're logged in as an admin, you can see the scheduled tasks that will run by going to /api/zhong on your instance. 4.1 added packaging tasks, which can be resource intensive. Packaging is used to generate EPUBs from Manifold texts and to create baggit packages of projects. That should only be running when projects change, so I wonder if something is happening to make Manifold's packaging too aggressive (ccing @scryptmouse, who worked on this feature). You might try disabling these packaging tasks temporarily, which I believe you can do from that Zhong management interface, and see if that mitigates the load spikes. If it does, at least we've narrowed down the problem.

**Comment by zdavis** [comment]
@bot bug

**Comment by zdavis** [comment]
Thank you for taking the time to report this bug. We've reviewed the issue and agree that this should be fixed. We’re adding a “planned” label to indicate that we consider this bug as part of our current work plan.

_This was an automated message, but please don't hesitate to reply. Our team watches these issues closely and will respond as soon as we're able to!_

**Comment by Cutuchiqueno** [comment]
`/opt/manifold/embedded/src/api` gives me:

```
require "./config/boot"
require "./config/environment"

Zhong.redis = Redis.new(url: ENV["RAILS_REDIS_URL"])

Zhong.schedule do
  category "social" do
    every(1.hour, "queue_fetch_project_tweets") { ::QueueFetchProjectTweets.perform_later }
  end

  category "caches" do
    every(15.minutes, "refresh_project_collections") do
      ::ProjectCollectionJobs::QueueCacheCollectionProjectsJob.perform_later
    end

    every(4.hours, "update_statistics") { ::UpdateAnalyticsCache.perform_later }
  end

  category "uploads" do
    every(1.day, "expire_shrine_cache", at: "22:00", tz: "America/Los_Angeles") do
      ExpireShrineCacheJob.perform_later
    end

    every(1.day, "expire_tus_uploads", at: "23:00", tz: "America/Los_Angeles") do
      ExpireTusUploadsJob.perform_later
    end
  end

  category "notification" do
    every(1.day, "enqueue_user_daily_digests", at: "06:00") do
      Notifications::EnqueueDigestsJob.perform_later "daily"
    end

    every(1.week, "enqueue_user_weekly_digests", at: "Sunday 06:00") do
      Notifications::EnqueueDigestsJob.perform_later "weekly"
    end
  end

  category "packaging" do
    every(5.minutes, "automate_text_exports") do
      Texts::AutomateExportsJob.perform_later
    end

    every(1.day, "prune_text_exports", at: "01:00") do
      TextExports::PruneJob.perform_later
    end

    every(4.hours, "prune_bag_it_temporary_directory") do
      Packaging::BagItSpec::PruneTemporaryDirectoryJob.perform_later
    end
  end
end
```

Comparing the jobs to the frequency of peaks in the diagram it is definitely the `ProjectCollectionJob` since it runs 4x times an hour and there are 4 peaks an hour.

Unfortunately I didn't manage to use zhong interface. Starting to use it from the server envrionment does not work because there is no ruby installed and using it with `/opt/manifold/embedded/src/api/bin/ruby`  results in:

```
Traceback (most recent call last):
        1: from zhong:12:in `<main>'
zhong:12:in `chdir': Not a directory @ dir_s_chdir - api (Errno::ENOTDIR)
```

Maybe you can tell me how to deactivate this - or other - rules exactly and then I can test if this cause the load?

**Comment by zdavis** [comment]
You should be able to see the interface in your browser:
![image](https://user-images.githubusercontent.com/255909/80647085-bd705c00-8a22-11ea-83f8-7e5803acab3e.png)

Just out of curiosity, is this causing actual problems or performance degradation? Seeing a very short CPU spike to 25+/-% when some computation isn happening doesn't actually seem unusual to me, especially with an application like Manifold that's doing a number of things.

best,
Zach

**Comment by Cutuchiqueno** [comment]
Unfortunately, zhong seems to have nothing to do with it as we guessed. I turned off any job that runs with "high" frequency but there is no difference for the load.

Until now, I did not experience any performance issues from a user experience point of view, so might be ok to ignore it. It just creates attention when a publication behaves significantly different performance-wise after an upgrade. Thus, the reason why I posted this was actually to figure out if this is something to worry about or to be expected together with the upgrade.
